include/dxc/DXIL/DxilFunctionProps.h

```diff
@@ -25,6 +25,93 @@ class Constant;
 } // namespace llvm
 
 namespace hlsl {
+
+// SM 6.6 allows WaveSize specification for only a single required size.
+// SM 6.8+ allows specification of WaveSize as a min, max and preferred value.
+struct DxilWaveSize {
+  unsigned Min = 0;
+  unsigned Max = 0;
+  unsigned Preferred = 0;
+
+  DxilWaveSize() = default;
+  DxilWaveSize(unsigned min, unsigned max = 0, unsigned preferred = 0)
+      : Min(min), Max(max), Preferred(preferred) {}
+  DxilWaveSize(const DxilWaveSize &other) = default;
+  DxilWaveSize &operator=(const DxilWaveSize &other) = default;
+  bool operator==(const DxilWaveSize &other) const {
+    return Min == other.Min && Max == other.Max && Preferred == other.Preferred;
+  }
+
+  // Encode valid DxilWaveSize, handling potential degenerate cases.
+  static DxilWaveSize Encode(unsigned min, unsigned max = 0,
+                             unsigned preferred = 0) {
+    if (max == min)
+      max = 0;
+    if (max == 0 && preferred == min)
+      preferred = 0;
+    return DxilWaveSize(min, max, preferred);
+  }
+
+  // Valid non-zero values are powers of 2 between 4 and 128, inclusive.
+  static bool IsValidValue(unsigned Value) {
+    return (Value >= 4 && Value <= 128 && ((Value & (Value - 1)) == 0));
+  }
+  // Valid encodings:
+  //  0, 0, 0: Not defined
+  //  Min, 0, 0: single WaveSize (SM 6.6/6.7)
+  //  Min, Max (> Min), 0 or Preferred (>= Min and <= Max): Range (SM 6.8+)
+  enum class ValidationResult {
+    Success,
+    InvalidMin,
+    InvalidMax,
+    InvalidPreferred,
+    MaxOrPreferredWhenUndefined,
+    PreferredWhenNoRange,
+    MaxEqualsMin,
+    MaxLessThanMin,
+    PreferredOutOfRange,
+  };
+  ValidationResult Validate() const {
+    if (Min == 0) { // Not defined
+      if (Max != 0 || Preferred != 0)
+        return ValidationResult::MaxOrPreferredWhenUndefined;
+    } else if (!IsValidValue(Min)) {
+      return ValidationResult::InvalidMin;
+    } else if (Max == 0) { // single WaveSize (SM 6.6/6.7)
+      if (Preferred != 0)
+        return ValidationResult::PreferredWhenNoRange;
+    } else if (!IsValidValue(Max)) {
+      return ValidationResult::InvalidMax;
+    } else if (Min == Max) {
+      return ValidationResult::MaxEqualsMin;
+    } else if (Max < Min) {
+      return ValidationResult::MaxLessThanMin;
+    } else if (Preferred != 0) {
+      if (!IsValidValue(Preferred))
+        return ValidationResult::InvalidPreferred;
+      if (Preferred < Min || Preferred > Max)
+        return ValidationResult::PreferredOutOfRange;
+    }
+    return ValidationResult::Success;
+  }
+  bool IsValid() const { return Validate() == ValidationResult::Success; }
+
+  bool IsDefined() const { return Min != 0; }
+  bool IsRange() const { return Max != 0; }
```

Comment:
Per the spec, this tag can only represent ranged wave sizes

Reply:
Nevermind, I wasn't clear on the use of this struct yet.

Reply:
No, this is not a "tag" this is a state structure that every entry has, which must be able to represent each possible state: no wave size defined, wave size with no range defined, and wave size with range defined.